src/libraries/System.DirectoryServices.Protocols/tests/DirectoryServicesProtocolsTests.cs

```diff
@@ -630,7 +628,12 @@ private LdapConnection GetConnection()
             // Set server protocol before bind; OpenLDAP servers default
             // to LDAP v2, which we do not support, and will return LDAP_PROTOCOL_ERROR
             connection.SessionOptions.ProtocolVersion = 3;
-            connection.SessionOptions.SecureSocketLayer = LdapConfiguration.Configuration.UseTls;
+
+            if (LdapConfiguration.Configuration.TransportSecurityOption == LdapConfiguration.TransportSecurity.StartTls)
+                connection.SessionOptions.StartTransportLayerSecurity(null);
+            else if (LdapConfiguration.Configuration.TransportSecurityOption == LdapConfiguration.TransportSecurity.Tls)
+                connection.SessionOptions.SecureSocketLayer = true;
+
             connection.Bind();
```

Comment:
I don't suppose you've been able to run a test like this from a Linux client against a Windows AD server?  In testing I've done this past week I've found that the concept worked in the 5.0 package, but it fails in weird ways with the 6.0/main build.

Reply:
Do you mean the StartTLS config, or the TLS config? What kind of issues have you run into?I have run the test suite with both of those configs against a local OpenLDAP instance, but I can't run the test suite against AD as I don't have a test instance of AD available

However, I have been using a locally built package from main to develop an application which uses an AD instance. (The `System.DirectoryServices.Protocols.6.0.0-preview.6.21352.12` package binds OK for me, but because of bogus referrals in our AD instance, I can't use that one yet. That issue has been fixed on `main`.) For now, I am using a package built from this PR branch, and everything (binding in cleartext, TLS and StartTLS, and enabling/disabling referrals) works as expected for my use case.

Reply:
The error I saw based on main as of ~2 weeks ago (https://github.com/dotnet/runtime/commit/dfdec3bfbb09a2d9d865651925b1679dd1f5aa7a or so) was that 

```C#
LdapConnection connection = new LdapConnection(host);
host.SessionOptions.ProtocolVersion = 3;
host.SessionOptions.StartTransportLayerSecurity(null);
```

got an error out of OpenLDAP without ever connecting.  I didn't finish tracking it down.

Reply:
Yeah, this is the issue that this PR is intended to fix.

The problem is that when I added TLS support in #52904, I deferred setting the URI on the OpenLDAP connection until `Bind()`, forgetting that StartTLS requires sending an LDAP operation to the host before binding. Since there's not a URI specified yet, `StartTransportLayerSecurity()` can't send the request anywhere and fails.

Reply:
For that same reason I would avoid delaying more things until Bind() is called. Seems like when adding support for ldaps and delaying the connection we accidentally broke the call to startTLS, I would rather not keep delaying more things (like the actual call to startTLS) as that might have other unintended consequences.